nomenclate/core/configurator.py

```diff
@@ -38,150 +25,214 @@ def __init__(self, path=None):
+        """From: http://stackoverflow.com/questions/13062300/convert-a-dict-to-sorted-dict-in-python
         """
-        query_data = self.data[section]
-        if isinstance(query_data, dict):
-            return [k for k in self.data[section].keys()]
-        elif isinstance(query_data, string_types):
-            return query_data.split(' ')
-        return query_data
+        try:
+            items = list(input.iteritems())
+        except AttributeError:
+            items = list(input.items())
+        return OrderedDict(sorted(items, key=lambda x: x[0]))
+
+
+class ListToStringEntryFormatter(BaseFormatter):
+    converts = {'takes': list,
+                'returns': str}
 
     @staticmethod
-    def _dict_to_ordered_dict(d):
-        """ Private function for converting to ordered dictionary
-            From: http://stackoverflow.com/questions/13062300/convert-a-dict-to-sorted-dict-in-python
+    def format_result(input):
+        return ' '.join(input)
+
+
+class ConfigEntryFormatter(object):
+    def format_query_result(self, query_result, query_path, return_type=list, preceding_depth=None):
+        """
         Args:
-            d (dictionary): unsorted dictionary
-        Returns (OrderedDictionary): dictionary in order...
+            query_result (dict|str|list): yaml query result
+            query_path [str]: list of strings representing query path
+            return_type (rtype): return type of object user desires
+            preceding_depth (int): the depth to which we want to encapsulate back up config tree
+                                    -1 : defaults to entire tree
+        Returns (dict|OrderedDict|str|list): specified return type
         """
+        if type(query_result) != return_type:
+            converted_result = self.format_with_handler(query_result, return_type)
+        else:
+            converted_result = query_result
+
+        converted_result =self.add_preceding_dict(converted_result, query_path, preceding_depth)
+        return converted_result
+
+    def get_handler(self, query_result, return_type):
         try:
-            items = list(d.iteritems())
+            return FormatterRegistry.get_by_take_and_return_type(query_result, return_type)
         except AttributeError:
-            items = list(d.items())
-        return OrderedDict(sorted(items, key=lambda x: x[0], reverse=True))
+            raise AttributeError(
+                "Handler not found for return type %s and input type %s" % (return_type, type(query_result)))
+        except:
+            raise IndexError('Could not find function in conversion list',
+                             'for input type %s and return type %s' % (query_result, return_type))
+
+    def format_with_handler(self, query_result, return_type):
+        handler = self.get_handler(type(query_result), return_type)
+        return handler.format_result(query_result)
+
+    def add_preceding_dict(self, config_entry, query_path, preceding_depth):
```

Comment:
I think it complicates it.  Mainly because I still don't really understand it and no function should be so complicated that it requires a few explanations.